Изменение проверки регистрации пользователя и сообщения об успешной авторизации.

eugemos opened the pull request:
### Что сделано
#### Проверка регистрации пользователя и обработка её отсутствия теперь работает так:
1. Если открыть бот прямо в Телеграме, то
    1. если в таблице User есть запись с соответствующим telegram_id (то есть с telegram_id текущего пользователя) , то
        1. если у этой записи external_id = null, то команды /start и /menu выводят сообщение с требованием регистрации;
        2. если у этой записи external_id != null, то по команде /start выводится сообщение об успешной авторизации;
    2. если в таблице User нет записи с соответствующим telegram_id, то команды /start и /menu выводят сообщение с требованием регистрации;

2. Если перейти в бот с помощью браузера по ссылке вида `https://t.me/<name_of_bot>?start=<id_hash>` и нажать кнопку `ЗАПУСТИТЬ` то
    1. если в таблице ExternalSiteUser есть запись с соответствующим id_hash, то
         1. если в таблице User есть запись, связанная с данной записью, и она имеет соответствующий telegram_id, то выводится сообщение об успешной авторизации;
         2. если в таблице User есть запись (№1), связанная с данной записью, и она имеет другой telegram_id, но имеется другая запись (№2), имеющая соответствующий telegram_id, то связь записи №1 с записью в таблице ExternalSiteUser разрывается (external_id = null), запись №2 связывается с записью из таблицы ExternalSiteUser  и обновляется данными из неё. Затем выводится сообщение об успешной авторизации;
         3. если в таблице User есть запись (№1), связанная с данной записью, но она имеет другой telegram_id, и нет другой записи, имеющей соответствующий telegram_id, то связь записи №1 с записью в таблице ExternalSiteUser разрывается (external_id = null), затем создаётся новая запись в таблице User с соответствующим  telegram_id, связывается с записью из таблицы ExternalSiteUser и обновляется данными из неё. Затем выводится сообщение об успешной авторизации;
         4. если в таблице User нет записи, связанной с данной записью, но имеется запись (№2), имеющая соответствующий telegram_id, то эта запись связывается с записью из таблицы ExternalSiteUser  и обновляется данными из неё. Затем выводится сообщение об успешной авторизации;
         5. если в таблице User нет записи, связанной с данной записью, и нет записи, имеющей соответствующий telegram_id, то создаётся новая запись с соответствующим telegram_id, она связывается с записью из таблицы ExternalSiteUser  и обновляется данными из неё. Затем выводится сообщение об успешной авторизации;
    2. если в таблице ExternalSiteUser отсутствует запись с соответствующим id_hash, то выводится сообщение с требованием регистрации.
 
#### В сообщение с требованием регистрации добавлена кнопка `Войти в личный  кабинет`

#### Сообщение об успешной авторизации изменено согласно условиям задачи #418

### Как проверял
Проверял локально. Соответствующие ситуации создавал путем ручного изменения БД через DBeaver. После ручного изменения БД бот надо перезапускать, иначе он ведёт себя непредсказуемо.

### Замечания
1. Алгоритм проверки регистрации и вывод сообщения с требованием зарегистрироваться вынесены в декоратор, что позволяет легко применить его и в других обработчиках без повторов кода. Алгоритма обработки отсутствия регистрации, приведённого выше, там нет, так как он нужен только в обработчике команды /start.
2. Строка 63 в файле `src/core/db/repository/user.py` - это костыль. Без него программа падает после создания новой записи в таблице User c сообщением `InvalidRequestError: A transaction is already begun on this Session.` и не обновляет категории. Почему так, мне не понятно. Может кто-нибудь подскажет.

eugemos committed:
Добавлен свойства для получения url-ссылок на страницу регистрации и
страницу авториации волонтёра.

eugemos committed:
В ExternalSiteUserService добавлен метод get_by_telegram_id

eugemos committed:
Изменена проверка регистрации пользователя для обработчика команды /start
в соответствии с условием задачи #412.

eugemos committed:
Алгоритм проверки регистрации пользователя и обработки
её отсутствия вынесен в декоратор.
Применено в обработчиках команд /start и /menu.

eugemos committed:
Корректировки в процессе отладки обработчика команды /start.

eugemos committed:
Сообщение об успешной регистрации изменено согласно условиям задачи #418.

eugemos committed:
Merge branch 'develop' into fix/change-user-auth-412

eugemos committed:
Merge branch 'develop' into fix/change-user-auth-412

gorskyolga commented on src/bot/handlers/decorators.py:
```diff
@@ -0,0 +1,46 @@
+                    "Чтобы получить доступ к боту, "
+                    "авторизуйтесь или зарегистрируйтесь на платформе."
+                ),
+                parse_mode="HTML",
```
Везде по коду используется parse_mode=ParseMode.HTML

gorskyolga commented on src/bot/handlers/registration.py:
```diff
@@ -6,46 +6,34 @@
-        "Изменить настройку уведомлений можно в личном кабинете.\n\n",
+        f'Изменить настройку уведомлений можно в <a href="{volunteer_auth_url}">личном кабинете</a>.\n\n'
+        "Навигация по боту запускается командой /menu.",
+        parse_mode="HTML",
```
Везде по коду используется parse_mode=ParseMode.HTML

gorskyolga replied:
В сообщении отображается превью ссылки. Нужно добавить disable_web_page_preview=True.

eugemos committed:
Merge branch 'develop' into fix/change-user-auth-412

eugemos committed:
Исправлено после ревью.

eugemos committed:
Merge branch 'develop' into fix/change-user-auth-412

gorskyolga commented on src/bot/keyboards.py:
```diff
@@ -14,12 +16,18 @@
+    volunteer_auth_url: str = Provide[Container.settings.provided.procharity_volunteer_auth_url],
+) -> list[InlineKeyboardButton]:
+    url = volunteer_auth_url if registered else registration_url
+    return [InlineKeyboardButton("🚪 Перейти в личный кабинет", url=url)]
```
Согласно схеме Миро, текст на кнопке "Войти в личный кабинет".

eugemos replied:
Упс-с, и вдобавок ещё и текст на кнопке в основном меню испортил. Какое-то наваждение. Тоже исправлю.

eugemos committed:
Исправлены ошибки в кнопках.

eugemos committed:
Merge branch 'develop' into fix/change-user-auth-412

eugemos committed:
Merge branch 'develop' into fix/change-user-auth-412

gorskyolga commented on src/core/db/repository/user.py:
```diff
@@ -60,6 +60,7 @@ async def update_bot_banned_status(self, user: User, banned: bool) -> None:
 
     async def set_categories_to_user(self, user_id: int, categories_ids: list[int]) -> Sequence[UsersCategories]:
         """Присваивает пользователю список категорий."""
+        await self._session.commit()
```
Закомментировала эту строку и не могу поймать ошибку `InvalidRequestError: A transaction is already begun on this Session.`. Она точно появляется после создания новой записи в таблице User?
Подробно я еще не тестировала разные случаи. Буду тестировать с закомментированной строкой.

eugemos replied:
Да. Это происходит, когда заходим в бот по ссылке с сайта. При этом, в таблице external_site_users соответствующая запись есть,  а в таблице users подходящей записи нет (D5 в твоей таблице). Запись в таблице users создаётся, а затем необходимо создать компетенции для нового пользователя - и вот при их создании это и происходит.
Попробовал тоже сейчас закомментировать - ошибка возникла.

gorskyolga commented on src/bot/handlers/decorators.py:
```diff
@@ -0,0 +1,47 @@
+from src.core.depends import Container
+
+
+def registered_user_required(handler):
```
Нужно добавить докстринг.

gorskyolga replied:
Может быть стоит перенести в src/bot/utils.py? Или это приводит к цикличным импортам?

eugemos committed:
Декоратор delete_previous_message перенесён в src/bot/utils.py,
к нему добавлен докстринг.

eugemos committed:
Merge branch 'develop' into fix/change-user-auth-412

gorskyolga commented on src/bot/handlers/registration.py:
```diff
@@ -2,51 +2,40 @@
-        )
-
+    telegram_user = update.effective_user
+    await user_service.register_or_update_user(ext_site_user, telegram_user)
```
Нужно сделать так, чтобы данные в users и user_categories обновлялись только при создании новой связи между юзером telegram и юзером внешнего сайта.
Соответственно в пунктах 1.i.b и 2.i.a данные не должны обновляться.
В рамках пункта 2.i.a не должна обрабатываться ситуация, когда пользователь внёс изменения в ЛК и применяет их в боте и поэтому необходимо обновить данные в таблицах user и user_categories данными из таблицы external_site_user. Этим уже занимается другая функция: при обновлении данных в ЛК вызывается эндпоинт `api/v1/auth/external_user_registration` и далее функция register из ExternalSiteUserService создает / обновляет юзера внешнего сайта и если есть связь с юзером телеграм, то обновляет у него категории. Может быть, если убрать это двойное обновление категорий, и костыль в 63 строке не понадобится...

eugemos committed:
По результатам ревью +
- метод register_or_update_user переименован;
- в нём немного упрощена логика;
- применил Never для однообразия.

eugemos committed:
Merge branch 'develop' into fix/change-user-auth-412

eugemos commented:
Но костыль по прежнему необходим.

gorskyolga commented on src/bot/services/user.py:
```diff
@@ -7,58 +9,50 @@ def __init__(self, user_repository: UserRepository, ext_user_repository: Externa
-        """Регистрирует нового пользователя по telegram_id.
-
-        Если пользователь найден, обновляет имя и флаг "заблокирован".
+    async def _update_or_create(self, user: User, **attrs) -> User:
```
Каждый раз смотрю на этот рефакторинг и радуюсь, как красиво переделано.

eugemos replied:
Не зря учили.

gorskyolga commented:
Для инфо:
По коду всё ок. Сегодня Халгина будет тестировать все возможные случаи. Судя по коду, должно всё верно отрабатывать.

eugemos committed:
Merge branch 'develop' into fix/change-user-auth-412

eugemos committed:
Merge branch 'develop' into fix/change-user-auth-412

gorskyolga committed:
Merge branch 'develop' into fix/change-user-auth-412

gorskyolga commented:
При тестировании обнаружили баг.
Как повторить баг:
1. Анна регистрируется в ЛК на сайте, заходит в бота и привязывает учётку Telegram к своему ЛК.
2. Борис заходит в бота по ссылке из ЛК Анны (с id_hash Анны). Учётка Telegram Бориса привязывается к ЛК Анны. Учётка Telegram Анны теряет связь с ЛК Анны.
3. Анна заходит в бота по ссылке из ЛК Анны (с id_hash Анны).
Ожидается: 
Учётка Telegram Анны привязывается к ЛК Анны. Учётка Telegram Бориса теряет связь с ЛК Анны.
В реальности:
Команды /start и /menu, как и должны, предлагают авторизоваться. Команда "/start <id_hash>" сообщает об успешной авторизации. При этом связь ЛК и учётки Telegram Анны не происходит. ЛК в БД по-прежнему связан с учёткой Telegram Бориса.

Примечания: для регистрации в ЛК локально использовали эндпоинт external_user_registration.

eugemos commented:
Сколько раз в описанном случае вы использовали эндпойнт external_user_registration: только один раз в п.1 или ещё где-то?

eugemos committed:
Merge branch 'develop' into fix/change-user-auth-412

gorskyolga commented:
> Сколько раз в описанном случае вы использовали эндпойнт external_user_registration: только один раз в п.1 или ещё где-то?

Только один раз как имитацию регистрации в ЛК.

eugemos commented:
У меня баг не воспроизводится.
Выполнял пункты 2 и 3 по кругу не один раз - всё норм, учётки переключаются.
Вы по-честному две разных учётки использовали или хитрили как-то?

gorskyolga commented:
> У меня баг не воспроизводится. Выполнял пункты 2 и 3 по кругу не один раз - всё норм, учётки переключаются. Вы по-честному две разных учётки использовали или хитрили как-то?

Да, честно использовали две разные учётки. Я сделаю запись экрана.

eugemos committed:
По результатам отладки.

eugemos commented:
Я нашёл строку, которая работала не всегда верно. Но причины этого не ясны, вроде, всё законно было. Как-то тут асинхронность нам боком выходит.
Заменил её на более надёжный вариант. Вроде, всё работает.

eugemos committed:
Добавил обновление поля role.
